### ISSUE TITLE
docker.service's DOCKER_OPTS is ignored when flanneld.service is active

### ISSUE THREAD

**Issue by tatsuya6502** [coreos/bugs#244]
Hello.

I'm trying CoreOS 557.0.0 from alpha channel.

```
core@core-01 ~ $ cat /etc/os-release
NAME=CoreOS
ID=coreos
VERSION=557.0.0
VERSION_ID=557.0.0
```

I'm running **flanneld.service**. And also I'd like to override `DOCKER_OPTS` in **docker.service** from cloud config. So my cloud config looks like this.

```
  units:
    - name: etcd.service
      command: start
    - name: flanneld.service
      # IMPORTANT: flanneld.service needs to be listed before any services that
      # run Docker containers. This includes fleet.service as fleet daemon will
      # start unit that may run Docker containers.
      # https://coreos.com/docs/cluster-management/setup/flannel-config/
      drop-ins:
        - name: 50-network-config.conf
          content: |
            [Service]
            ExecStartPre=/usr/bin/etcdctl set /coreos.com/network/config '{ "Network": "10.1.0.0/16" }'
      command: start
    - name: fleet.service
      command: start
    - name: docker.service
      drop-ins:
        - name: 50-dns-search.conf
          content: |
            [Service]
            Environment='DOCKER_OPTS=--dns-search=localdomain --insecure-registry="0.0.0.0/0"'
```

However, my `DOCKER_OPTS` doesn't appear in systemd status (the last line for PID 2456).

```
core@core-01 ~ $ cat /etc/systemd/system/docker.service.d/50-dns-search.conf
[Service]
Environment='DOCKER_OPTS=--dns-search=localdomain --insecure-registry="0.0.0.0/0"'

core@core-01 ~ $ sudo systemctl start docker.service
core@core-01 ~ $ systemctl status docker.service
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib64/systemd/system/docker.service; disabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─50-dns-search.conf
   Active: active (running) since Fri 2015-01-16 10:19:55 UTC; 6s ago
     Docs: http://docs.docker.com
 Main PID: 2456 (docker)
   CGroup: /system.slice/docker.service
           └─2456 docker --daemon --host=fd:// --ip-masq=false --bip=10.1.13.1/24 --mtu=1472
```

I think my `DOCKER_OPTS` is overwritten by another `DOCKER_OPTS` in `/run/docker_opts.env`. This file is created by flanneld.service.

Here is a relevant section from the doc.

https://coreos.com/docs/cluster-management/setup/flannel-config/

> - 5. `ExecStartPost` in `flanneld.service` converts information in
>   `/run/flannel/subnet.env` into Docker daemon command line args (such
>   as `--bip` and `--mtu`), storing them in `/run/docker_opts.env`
> - ...
> - 7. `docker.service` sources in `/run/docker_opts.env` (which
>   contains `DOCKER_OPTS`) and starts the Docker daemon with
>   `$DOCKER_OPTS`

I think we cloud rename this `DOCKER_OPTS` (from flanneld.service) to something else like `DOCKER_NET_OPTS`, so that we can have the both options in effect?

Thanks,

**Comment by marineam** [comment]
Yeah, we need to add a new $FLANNEL_DOCKER_OPTS or some such. Also /run/docker_opts.env should be renamed to indicate that it for flannel.

**Comment by eyakubovich** [comment]
@marineam It was my intention that /run/docker_opts.env (and DOCKER_OPTS) could contain any extra docker options, not just flannel related. For example, as early-docker services execute, they could add to the DOCKER_OPTS. But since we want DOCKER_OPTS to be user over-riddable, maybe we should call env var in /run/docker_opts.env SYS_DOCKER_OPTS.

At the same time, once https://github.com/coreos/flannel/pull/93 gets merged, I was thinking of just using individual opts - DOCKER_OPT_BIP & DOCKER_OPT_MTU.

**Comment by msumme** [comment]
I'm running into this issue as well.  Is there a workaround?  I've tried a few different approaches, and so far have been unsuccessful in getting docker to load with both flannel options and my own options (insecure-registry for local environment)

**Comment by tatsuya6502** [comment]
@msumme -

I'm currently running the following command from Ansible to add my option to `/run/docker_opts.env`.

```
sed -i.org 's/DOCKER_OPTS="/DOCKER_OPTS="--dns-search=localdomain /g' /run/docker_opts.env
```

But it has an obvious problem. It won't survive when flanneld.service is restarted. This is ok to me as I launch my CoreOS cluster on EC2 from Jenkins and destroy it after finishing some tests. My cluster will live only for few hours.

Another way would be to completely replace docker.service with your own version of unit file.

1. Copy the original unit file for docker.service to `/etc/systemd/system/`
  You'll find the path to the original unit file by running `systemctl status docker.service`
2. Edit your unit file as you want.
3. Run the following commands to enable your unit file.

```
cd /etc/systemd/system/
systemctl enable docker.service
systemctl start docker.service
```

I haven't tried, but I think you can do this via cloud config.

**Comment by crawford** [comment]
This has been fixed in 557.2.0 (currently rolling to Beta) and should soon be deployed to Alpha as well.
